### PR TITLE
Add spring oauth2 resource server to classpath and add anonymous test autoconfiguration

### PR DESCRIPTION
**Pull request by NielsCW**

### CODE REVIEW

File: contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousTestAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.security.web.authentication.AnonymousAuthenticationFilter;
+
+@AutoConfiguration(before = { OAuth2ResourceServerAutoConfiguration.class, SecurityAutoConfiguration.class })
+@ConditionalOnClass({SecurityFilterChain.class, HttpSecurity.class})
+@ConditionalOnProperty("contentgrid.spring.security.allow-anonymous")
+@EnableWebSecurity
+public class AnonymousTestAutoConfiguration {
+
+    @Bean
+    @ConditionalOnClass(BearerTokenAuthenticationToken.class)
```
Comment (tgeens) [inline contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousTestAutoConfiguration.java:31]: add `@ConditionalOnDefaultWebSecurity`

File: contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousTestAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.server.resource.authentication.BearerTokenAuthenticationToken;
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
+import org.springframework.security.web.SecurityFilterChain;
+import org.springframework.security.web.authentication.AnonymousAuthenticationFilter;
+
+@AutoConfiguration(before = { OAuth2ResourceServerAutoConfiguration.class, SecurityAutoConfiguration.class })
+@ConditionalOnClass({SecurityFilterChain.class, HttpSecurity.class})
+@ConditionalOnProperty("contentgrid.spring.security.allow-anonymous")
```
[the remark this answers is not in the document]
Comment (tgeens) [reply contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousTestAutoConfiguration.java:26]:
```
boolean matchIfMissing() default false;
```

File: contentgrid-spring-data-support/src/main/java/com/contentgrid/spring/data/support/auditing/v1/JwtAuditorAware.java [hunk 3]
```diff
@@ -15,11 +15,17 @@ public Optional<UserMetadata> getCurrentAuditor() {
...
+                .map(principal -> {
+                    if (principal instanceof JwtClaimAccessor jwt) {
+                        if (jwt.getSubject() != null && jwt.getIssuer() != null) {
+                            return new UserMetadata(jwt.getSubject(), jwt.getIssuer().toString(),
+                                    jwt.hasClaim("name") ? jwt.getClaimAsString("name") : jwt.getSubject());
+                        }
+                    } else if (principal instanceof String name) {
+                        // Anonymous user when 'contentgrid.security.unauthenticated.allow' is true
+                        return new UserMetadata(null, null, name);
+                    }
```
Comment (vierbergenlars) [inline contentgrid-spring-data-support/src/main/java/com/contentgrid/spring/data/support/auditing/v1/JwtAuditorAware.java:27]: It seems a bit dangerous to assume any principal that is a string is an unauthenticated principal?

On the other hand; if the principal is a string, we can't really get more information about the user.

The `Authentication` itself does have a `getName()` method that would return the "name" (the id/subject); which we would use in here?

Lacking more information for an anonymous user; that would mean that the subject is set to `anonymousUser`; namespace would be `null` and the display name would also be `anonymousUser`?

File: contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousHttpConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.security.core.authority.AuthorityUtils;
+import org.springframework.security.web.authentication.AnonymousAuthenticationFilter;
+
+public class AnonymousHttpConfigurer extends AbstractHttpConfigurer<AnonymousHttpConfigurer, HttpSecurity> {
+
+    @Override
+    public void init(HttpSecurity http) throws Exception {
+        ApplicationContext context = http.getSharedObject(ApplicationContext.class);
+        var disableCsrf = context.getEnvironment().getProperty("contentgrid.security.csrf.disabled", Boolean.class);
+        var allowUnauthenticated = context.getEnvironment().getProperty("contentgrid.security.unauthenticated.allow", Boolean.class);
```
Comment (tgeens) [inline contentgrid-spring-boot-autoconfigure/src/main/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousHttpConfigurer.java:19]: I would re-order this, but that might be personal preference

File: contentgrid-spring-boot-autoconfigure/src/test/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousHttpConfigurerTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                    assertThat(context).hasNotFailed();
+                    assertThat(context).hasSingleBean(SecurityFilterChain.class);
+                    assertThat(context.getBean(SecurityFilterChain.class).getFilters()).anyMatch(filter ->
+                            filter instanceof AnonymousUsernamePasswordAuthenticationFilter
+                    );
+                });
+    }
+
+    @Test
+    void checkWithoutOauth2ResourceServer() {
```
Comment (tgeens) [inline contentgrid-spring-boot-autoconfigure/src/test/java/com/contentgrid/spring/boot/autoconfigure/security/AnonymousHttpConfigurerTest.java:53]:
```suggestion
    void checkWithoutOAuth2ResourceServer() {
```

**Review by tgeens — APPROVED** [review]

**Pull request merged**